File: matlab/src/api/snc/snc_srv_rate.m
%{ @file snc_srv_rate.m
 %  @brief MGF service envelope of a constant-rate work-conserving server
 %
 %  @author LINE Development Team
%}

%{
 % @brief MGF service envelope of a constant-rate work-conserving server
 %
 % @details
 % A work-conserving server of capacity C offers the deterministic service
 % process S(s,t) = C*(t-s), whose exponential-form envelope
 %
 %   E[exp(-theta*S(s,t))] <= exp(-theta*(rho*(t-s) - sigma))
 %
 % holds with rho(theta) = C and sigma(theta) = 0 for every theta > 0. This is
 % the elementary service element of the snc family; a station shared by cross
 % traffic is obtained from it through snc_leftover, and a tandem of stations
 % through snc_conv.
 %
 % @par Syntax:
 % @code
 % [sigma,rho] = snc_srv_rate(C)
 % [sigma,rho] = snc_srv_rate(C,theta)   % theta ignored, kept for handle
 %                                       % compatibility
 % srv = @(theta) snc_srv_rate(C,theta);
 % @endcode
 %
 % @par Parameters:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>C<td>Server capacity, work per slot
 % <tr><td>theta<td>Optional Chernoff parameter, unused
 % </table>
 %
 % @par Returns:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>sigma<td>Burst term of the service envelope (0 here)
 % <tr><td>rho<td>Rate term of the service envelope, equal to C
 % </table>
 %
 % @par Reference:
 % M. Fidler and A. Rizk, "A Guide to the Stochastic Network Calculus",
 % IEEE Communications Surveys and Tutorials, 17(1):92-105, 2015, Sec. IV.
%}
function [sigma, rho] = snc_srv_rate(C, theta)

if nargin < 1
    line_error(mfilename, 'Usage: [sigma,rho] = snc_srv_rate(C).');
end
if C <= 0
    line_error(mfilename, 'C must be positive. Got %g.', C);
end
if nargin >= 2 && theta <= 0
    line_error(mfilename, 'theta must be positive. Got %g.', theta);
end

sigma = 0;
rho = C;
end

File: jar/src/main/java/jline/api/snc/Snc_srv_rate.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.api.snc;

/**
 * MGF service envelope of a constant-rate work-conserving server.
 *
 * <p>{@code S(s,t) = C*(t-s)}, so {@code (sigma, rho) = (0, C)} for every theta.
 * This is the elementary service element; a station shared by cross traffic is
 * obtained from it through {@link Snc_leftover} and a tandem through
 * {@link Snc_conv}.</p>
 *
 * <p>USE {@link Snc_srv_exp} INSTEAD WHENEVER THE WORK UNIT IS THE JOB: pairing
 * this element with a job-counting arrival envelope models an M/D/1 and
 * understates the delay of an exponential server.</p>
 *
 * <p>Port of matlab/src/api/snc/snc_srv_rate.m.</p>
 */
public final class Snc_srv_rate {
    private Snc_srv_rate() {}

    /**
     * @param C server capacity, work per slot
     * @return {0, C}
     */
    public static double[] snc_srv_rate(double C) {
        if (C <= 0) {
            throw new IllegalArgumentException("snc_srv_rate: C must be positive, got " + C);
        }
        return new double[] {0.0, C};
    }

    /**
     * @param C     server capacity
     * @param theta Chernoff parameter, accepted and ignored
     * @return {0, C}
     */
    public static double[] snc_srv_rate(double C, double theta) {
        if (theta <= 0) {
            throw new IllegalArgumentException("snc_srv_rate: theta must be positive, got " + theta);
        }
        return snc_srv_rate(C);
    }

    /**
     * @param C server capacity
     * @return the envelope as a function of theta
     */
    public static SncEnvelope of(final double C) {
        return new SncEnvelope() {
            public double[] eval(double theta) {
                return snc_srv_rate(C, theta);
            }
        };
    }
}
